udapi/block/tokenize/onwhitespace.py

```diff
@@ -1,36 +1,72 @@
 """Block tokenize.OnWhitespace"""
+import re
 from udapi.core.block import Block
 
 
 class OnWhitespace(Block):
-    """"Base tokenizer, splits on whitespaces, fills SpaceAfter=No."""
+    """Base tokenizer, splits on whitespaces, fills SpaceAfter=No.
+
+    Use the parameter `normalize_spaces=False` to preserve all whitespaces in the sentence
+    in the UDPipe way, i.e. using the `SpacesAfter` and `SpacesBefore` features in the MISC field.
+    It is backward compatible with CoNLL-U v2 `SpaceAfter=No` feature. That is, no following
+    whitespace is marked by `SpaceAfter=No` and a single following space results in no
+    whitespace-related markup.
+    If loading the text using `read.Sentences` and all whitespaces need to be preserved
+    (in order to be able to reconstruct the original document), the `read.Sentences` block
+    must be called with `rstrip=\n` or `rstrip=\r\n` to prevent stripping the trailing
+    whitespace, e.g.::
+        $> echo -e "Hello \t world " | udapy read.Sentences $'rstrip=\r\n' tokenize.OnWhitespace normalize_spaces=0 write.Conllu
+
+        # sent_id = 1
+        # text = Hello   world
+        1       Hello   _       _       _       _       0       _       _       SpacesAfter=\s\t\s
+        2       world   _       _       _       _       0       _       _       _
+    Note that the attribute `SpaceAfter=No` is missing for the token `world`, since it is
+    followed by a single space.
+
+    Parameters
+    ----------
+    normalize_spaces : bool
+        preserve whitespaces by filling MISC attributes `SpacesAfter` and `SpacesBefore` (by default True)
+    """
+
+    def __init__(self, normalize_spaces=True, **kwargs):
+        super().__init__(**kwargs)
+        self.normalize_spaces = normalize_spaces
 
     @staticmethod
     def tokenize_sentence(string):
         """A method to be overriden in subclasses."""
         return string.split()
 
+    @staticmethod
+    def escape_whitespace(string):
+        string = re.sub(r' ', r'\\s', string)
+        string = re.sub(r'\t', r'\\t', string)
+        string = re.sub(r'\r', r'\\r', string)
+        string = re.sub(r'\n', r'\\n', string)
+        return string
```

Comment:
Even better and faster would be to define a single global constant
`escape_whitespace_table = str.maketrans({' ':r'\s', '\t':r'\t', '\r':r'\r', '\n':r'\n'})`
and then substitute
`escape_whitespace(s)`
with
`s.translate(escape_whitespace_table)`

Reply:
OK, I'll do it this way.